File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/ZoneHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.str
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject

class ZoneHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.ZONE_INFO,
        RequestTypes.ZONE_LIST,
        RequestTypes.ZONE_TRAVEL,
        RequestTypes.ZONE_ATTACK,
        RequestTypes.ZONE_CLAIM,
        RequestTypes.FACTION_CYCLE_INFO
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.ZONE_INFO -> zoneInfo(ctx, payload)
            RequestTypes.ZONE_LIST -> zoneList(ctx)
            RequestTypes.ZONE_TRAVEL -> zoneTravel(ctx, playerId, payload)
            RequestTypes.ZONE_ATTACK -> zoneAttack(ctx, playerId, payload)
            RequestTypes.ZONE_CLAIM -> zoneClaim(ctx, playerId, payload)
            RequestTypes.FACTION_CYCLE_INFO -> factionCycle(ctx)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported zone request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun zoneInfo(ctx: HandlerContext, payload: JsonObject): HandlerResult {
        val zoneId = payload.str("zoneId")
            ?: return HandlerResult.Err(ServerError.fromMessage("zoneId required", ErrorCategory.VALIDATION))
        val snapshot = ctx.services.worldZoneService.getZoneSnapshot(zoneId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Zone not found", ErrorCategory.NOT_FOUND))
        return HandlerResult.Ok(ResponseTypes.ZONE_INFO, mapOf("zone" to snapshot))
    }

    private suspend fun zoneList(ctx: HandlerContext): HandlerResult {
        val zones = ctx.services.worldZoneService.getZoneSnapshots()
        return HandlerResult.Ok(ResponseTypes.ZONE_LIST, mapOf("zones" to zones))
    }

    private suspend fun zoneTravel(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val targetZoneId = payload.str("zoneId")
            ?: return HandlerResult.Err(ServerError.fromMessage("zoneId required", ErrorCategory.VALIDATION))
        val result = ctx.services.worldZoneService.initiateTravel(playerId, targetZoneId)
        return if (result.success) {
            HandlerResult.Ok(ResponseTypes.ZONE_TRAVEL_STARTED, mapOf(
                "message" to result.message,
                "travelTimeMs" to result.travelTimeMs
            ))
        } else {
            HandlerResult.Err(ServerError.fromMessage(result.message, ErrorCategory.VALIDATION))
        }
    }

    private suspend fun zoneAttack(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val targetZoneId = payload.str("zoneId")
            ?: return HandlerResult.Err(ServerError.fromMessage("zoneId required", ErrorCategory.VALIDATION))
        val result = ctx.services.worldZoneService.attackZone(playerId, targetZoneId)
        return if (result.success) {
            HandlerResult.Ok(ResponseTypes.ZONE_ATTACK_RESULT, mapOf(
                "message" to result.message,
                "threatGained" to result.threatGained,
                "repGained" to result.repGained
            ))
        } else {
            HandlerResult.Err(ServerError.fromMessage(result.message, ErrorCategory.VALIDATION))
        }
    }

    private suspend fun zoneClaim(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val targetZoneId = payload.str("zoneId")
            ?: return HandlerResult.Err(ServerError.fromMessage("zoneId required", ErrorCategory.VALIDATION))
        val result = ctx.services.worldZoneService.claimZone(playerId, targetZoneId)
        return if (result.success) {
            HandlerResult.Ok(ResponseTypes.ZONE_CLAIMED, mapOf(
                "message" to result.message
            ))
        } else {
            HandlerResult.Err(ServerError.fromMessage(result.message, ErrorCategory.VALIDATION))
        }
    }

    private suspend fun factionCycle(ctx: HandlerContext): HandlerResult {
        val cycle = ctx.services.worldZoneService.getActiveFactionCycle()
        return HandlerResult.Ok(ResponseTypes.FACTION_CYCLE, cycle)
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/ZoneHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class ZoneHandler : IHandler
{
    private readonly WorldZoneService _worldZoneService;
    private readonly IPlayerService _playerService;

    public string MessageType => "zone_info";

    public ZoneHandler(WorldZoneService worldZoneService, IPlayerService playerService)
    {
        _worldZoneService = worldZoneService;
        _playerService = playerService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        var zoneId = payload?.Str("zoneId");
        if (string.IsNullOrEmpty(zoneId))
            return new MessageResult("error", connectionId, new { message = "zoneId required" });

        var snapshot = await _worldZoneService.GetZoneSnapshot(zoneId);
        if (snapshot is null)
            return new MessageResult("error", connectionId, new { message = "Zone not found" });

        return new MessageResult("zone_info", connectionId, new { zone = snapshot });
    }

    public async Task<IActionResult> HandleZoneList(string connectionId, JsonElement? payload, string? requestId)
    {
        var zones = await _worldZoneService.GetZoneSnapshots();
        return new MessageResult("zone_list", connectionId, new { zones });
    }

    public async Task<IActionResult> HandleZoneTravel(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var zoneId = payload?.Str("zoneId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(zoneId))
            return new MessageResult("error", connectionId, new { message = "zoneId required" });

        var result = await _worldZoneService.InitiateTravel(playerId, zoneId);
        if (result.Success)
            return new MessageResult("zone_travel_started", connectionId, new
            {
                message = result.Message,
                travelTimeMs = result.TravelTimeMs
            });
        return new MessageResult("error", connectionId, new { message = result.Message });
    }

    public async Task<IActionResult> HandleZoneAttack(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var zoneId = payload?.Str("zoneId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(zoneId))
            return new MessageResult("error", connectionId, new { message = "zoneId required" });

        var result = await _worldZoneService.AttackZone(playerId, zoneId);
        if (result.Success)
            return new MessageResult("zone_attack_result", connectionId, new
            {
                message = result.Message,
                threatGained = result.ThreatGained,
                repGained = result.RepGained
            });
        return new MessageResult("error", connectionId, new { message = result.Message });
    }

    public async Task<IActionResult> HandleZoneClaim(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var zoneId = payload?.Str("zoneId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(zoneId))
            return new MessageResult("error", connectionId, new { message = "zoneId required" });

        var result = await _worldZoneService.ClaimZone(playerId, zoneId);
        if (result.Success)
            return new MessageResult("zone_claimed", connectionId, new { message = result.Message });
        return new MessageResult("error", connectionId, new { message = result.Message });
    }

    public IActionResult HandleFactionCycle(string connectionId, JsonElement? payload, string? requestId)
    {
        var cycle = _worldZoneService.GetActiveFactionCycle();
        return new MessageResult("faction_cycle", connectionId, cycle);
    }
}
